### ISSUE TITLE
Allow default binary protocol run in AOT environment with no code generation (.NET Standard 2.0 with no Emit)

### ISSUE THREAD

**Issue by dzmitry-lahoda** [aspnet/SignalR#3122]
I have tried to use `MessagePackHubProtocol` in Unity 2018.2(.NET Standrad 2.0) iOS and failed. 
Because `MessagePackHubProtocol` type loads `SignalRResolver` which type loads `emit` types which do not exists. Crash. 

I hope to have `MessagePackHubProtocol` to lazy load `emit` related types only if these are configured.

![img_20181012_120405](https://user-images.githubusercontent.com/757125/46859785-68a4ca80-ce17-11e8-97cb-12ee4e23cd3e.jpg)

```csharp
       public HubConnection Build(string serverPortHub/*, TODO: pass statically generated resolvers here */)
        {
            var transport = new HttpConnectionOptions()
            {
                Url = new Uri(serverPortHub),
                Transports = HttpTransportType.WebSockets
            };

            var serialization = new MessagePackHubProtocolOptions
            {
                FormatterResolvers = new List<IFormatterResolver>
                {
                    // add generated resolvers for our objects
                    
                    //https://github.com/neuecc/MessagePack-CSharp#pre-code-generationunityxamarin-supports
                    MessagePack.Resolvers.BuiltinResolver.Instance,
                    AttributeFormatterResolver.Instance,
                    MessagePack.Resolvers.PrimitiveObjectResolver.Instance,
                    //MessagePack.Unity.UnityResolver.Instance
                    //MessagePack.Resolvers.UnsafeBinaryResolver.Instances
                }
            };

            return new HubConnection(
                new HttpConnectionFactory(Options.Create(transport), _loggerFactory),
                new MessagePackHubProtocol(Options.Create(serialization)), _loggerFactory);
        }
```

**Comment by dzmitry-lahoda** [comment]
Removing usage of `SignalRResolver` is not enought as public APIs use internal `BinaryMessageFormatter` and `MemoryBufferWriter`.  I use https://www.nuget.org/packages/Microsoft.AspNetCore.SignalR.Protocols.MessagePack/1.1.0-preview2-35157

**Comment by dzmitry-lahoda** [comment]
And I have to copy paste
```
    public class MessagePackHubProtocolOptions
    {
        public IList<IFormatterResolver> FormatterResolvers { get; set; } = MessagePackHubProtocol.CreateDefaultFormatterResolvers();
}
```

**Comment by anurse** [comment]
Yep, there are some known issues with using the binary protocol where reflection emit isn't supported. This is an area we'll look at in 3.0.

**Comment by joshpeterson** [comment]
@dzmitry-lahoda:

Nothing from the exception message with IL2CPP initially jumps out at me as a cause. This looks like a normal .NET exception, so something earlier must be the problem. IL2CPP does not support managed code debugging, it is possible for you to debug this issue to better understand what the difference is between the Mono and IL2CPP cases?

**Comment by joshpeterson** [comment]
@dzmitry-lahoda: Thanks for the details, I'm glad that you have a work around with JSON.NET. I'll be happy to hear about any other developments from you or others.

**Comment by dzmitry-lahoda** [comment]
Here are some logs from XCode for code I have shared. I very new to Unity and iOS. So only high level logs, for exception screenshot attached earlier.

```
2018-10-15 16:42:47.326 TestTestTest[673:1260543] Built from '2018.3/staging' branch, Version '2018.3.0b5 (01088ee0a3a8)', Build type 'Release', Scripting Backend 'il2cpp'

2018-10-15 16:42:47.344 TestTestTest[673:1260543] -> registered mono modules 0x109d890

-> applicationDidFinishLaunching()

2018-10-15 16:42:49.006 TestTestTest[673:1260543] Metal GPU Frame Capture Enabled

2018-10-15 16:42:49.006 TestTestTest[673:1260543] Metal API Validation Disabled

-> applicationDidBecomeActive()

Renderer: PowerVR SGX 543

Vendor:   Imagination Technologies

Version:  OpenGL ES 2.0 IMGSGX543-124.1

GLES:     2

 GL_OES_depth_texture GL_OES_depth24 GL_OES_element_index_uint GL_OES_fbo_render_mipmap GL_OES_mapbuffer GL_OES_packed_depth_stencil GL_OES_rgb8_rgba8 GL_OES_standard_derivatives GL_OES_texture_float GL_OES_texture_half_float GL_OES_texture_half_float_linear GL_OES_vertex_array_object GL_EXT_blend_minmax GL_EXT_color_buffer_half_float GL_EXT_debug_label GL_EXT_debug_marker GL_EXT_discard_framebuffer GL_EXT_draw_instanced GL_EXT_instanced_arrays GL_EXT_map_buffer_range GL_EXT_occlusion_query_boolean GL_EXT_pvrtc_sRGB GL_EXT_read_format_bgra GL_EXT_separate_shader_objects GL_EXT_shader_framebuffer_fetch GL_EXT_shader_texture_lod GL_EXT_shadow_samplers GL_EXT_sRGB GL_EXT_texture_filter_anisotropic GL_EXT_texture_rg GL_EXT_texture_storage GL_APPLE_clip_distance GL_APPLE_color_buffer_packed_float GL_APPLE_copy_texture_levels GL_APPLE_framebuffer_multisample GL_APPLE_rgb_422 GL_APPLE_sync GL_APPLE_texture_format_BGRA8888 GL_APPLE_texture_max_level GL_APPLE_texture_packed_float GL_IMG_read_format GL_IMG_texture_com

pression_pvrtc

OPENGL LOG: Creating OpenGL ES 2.0 graphics device ; Context level  <OpenGL ES 2.0> ; Context handle 392715648

Initialize engine version: 2018.3.0b5 (01088ee0a3a8)

UnloadTime: 0.271375 ms

Debug RegisteringHandler Registering handler for client method 'Bye'.  Registering handler for client method 'Bye'.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace WaitingOnConnectionLock Waiting on Connection Lock in StartAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:289).  Waiting on Connection Lock in StartAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:289).

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug Starting Starting HubConnection.  Starting HubConnection.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug Starting Starting HttpConnection.  Starting HttpConnection.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug EstablishingConnection Establishing connection with server at 'http://192.168.100.3:5000/HelloWorld'.  Establishing connection with server at 'http://192.168.100.3:5000/HelloWorld'.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace SendingHttpRequest Sending HTTP request POST 'http://192.168.100.3:5000/HelloWorld/negotiate'.  Sending HTTP request POST 'http://192.168.100.3:5000/HelloWorld/negotiate'.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug Established Established connection 'AJvKxKmik6m3yNxakwnJng' with the server.  Established connection 'AJvKxKmik6m3yNxakwnJng' with the server.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug StartingTransport Starting transport 'WebSockets' with Url: http://192.168.100.3:5000/HelloWorld?id=AJvKxKmik6m3yNxakwnJng.  Starting transport 'WebSockets' with Url: http://192.168.100.3:5000/HelloWorld?id=AJvKxKmik6m3yNxakwnJng.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Information StartTransport Starting transport. Transfer mode: Binary. Url: 'ws://192.168.100.3:5000/HelloWorld?id=AJvKxKmik6m3yNxakwnJng'.  Starting transport. Transfer mode: Binary. Url: 'ws://192.168.100.3:5000/HelloWorld?id=AJvKxKmik6m3yNxakwnJng'.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


ERROR: 0:32: Use of undeclared identifier 'gl_FragDepthEXT'


Note: Creation of internal variant of shader 'Hidden/Internal-MotionVectors' failed.

WARNING: Shader Unsupported: 'Hidden/Internal-MotionVectors' - Pass '' has no vertex shader

WARNING: Shader Unsupported: 'Hidden/Internal-MotionVectors' - Setting to default shader.

Debug StartedTransport Started transport.  Started transport.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug TransportStarted Transport 'WebSockets' started.  Transport 'WebSockets' started.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Information Started HttpConnection Started.  HttpConnection Started.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Information HubProtocol Using HubProtocol 'messagepack v1'.  Using HubProtocol 'messagepack v1'.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug SendingHubHandshake Sending Hub Handshake.  Sending Hub Handshake.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug ReceivedFromApp Received message from application. Payload size: 39.  Received message from application. Payload size: 39.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug MessageReceived Message received. Type: Binary, size: 3, EndOfMessage: True.  Message received. Type: Binary, size: 3, EndOfMessage: True.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug HandshakeComplete Handshake with server complete.  Handshake with server complete.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug ReceiveLoopStarting Receive loop starting.  Receive loop starting.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug SendingMessageGeneric Sending PingMessage message.  Sending PingMessage message.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug MessageSentGeneric Sending PingMessage message completed.  Sending PingMessage message completed.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Information Started HubConnection started.  HubConnection started.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace ReleasingConnectionLock Releasing Connection Lock in StartAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:333).  Releasing Connection Lock in StartAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:333).

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug ReceivedFromApp Received message from application. Payload size: 3.  Received message from application. Payload size: 3.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace WaitingOnConnectionLock Waiting on Connection Lock in InvokeCoreAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:445).  Waiting on Connection Lock in InvokeCoreAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:445).

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace InvocationCreated Invocation 1 created.  Invocation 1 created.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace PreparingBlockingInvocation Preparing blocking invocation '1' of 'Hello', with return type 'System.String' and 1 argument(s).  Preparing blocking invocation '1' of 'Hello', with return type 'System.String' and 1 argument(s).

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug RegisteringInvocation Registering Invocation ID '1' for tracking.  Registering Invocation ID '1' for tracking.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace IssuingInvocation Issuing Invocation '1': System.String Hello(System.String).  Issuing Invocation '1': System.String Hello(System.String).

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Error FailedToSendInvocation Sending Invocation '1' failed. Sequence contains no elements Sending Invocation '1' failed.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace InvocationFailed Invocation 1 marked as failed.  Invocation 1 marked as failed.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace ReleasingConnectionLock Releasing Connection Lock in InvokeCoreAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:458).  Releasing Connection Lock in InvokeCoreAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:458).

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace WaitingOnConnectionLock Waiting on Connection Lock in InvokeCoreAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:445).  Waiting on Connection Lock in InvokeCoreAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:445).

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace InvocationCreated Invocation 2 created.  Invocation 2 created.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace PreparingBlockingInvocation Preparing blocking invocation '2' of 'Hello', with return type 'System.String' and 1 argument(s).  Preparing blocking invocation '2' of 'Hello', with return type 'System.String' and 1 argument(s).

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug RegisteringInvocation Registering Invocation ID '2' for tracking.  Registering Invocation ID '2' for tracking.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace IssuingInvocation Issuing Invocation '2': System.String Hello(System.String).  Issuing Invocation '2': System.String Hello(System.String).

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Error FailedToSendInvocation Sending Invocation '2' failed. Sequence contains no elements Sending Invocation '2' failed.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace InvocationFailed Invocation 2 marked as failed.  Invocation 2 marked as failed.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace ReleasingConnectionLock Releasing Connection Lock in InvokeCoreAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:458).  Releasing Connection Lock in InvokeCoreAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:458).

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace WaitingOnConnectionLock Waiting on Connection Lock in InvokeCoreAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:445).  Waiting on Connection Lock in InvokeCoreAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:445).

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace InvocationCreated Invocation 3 created.  Invocation 3 created.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace PreparingBlockingInvocation Preparing blocking invocation '3' of 'Hello', with return type 'System.String' and 1 argument(s).  Preparing blocking invocation '3' of 'Hello', with return type 'System.String' and 1 argument(s).

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Debug RegisteringInvocation Registering Invocation ID '3' for tracking.  Registering Invocation ID '3' for tracking.

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)


Trace IssuingInvocation Issuing Invocation '3': System.String Hello(System.String).  Issuing Invocation '3': System.String Hello(System.String).

Stacktrace is not supported on this platform. 

(Filename: ./Runtime/Export/Debug.bindings.h Line: 45)
```

**Comment by joshpeterson** [comment]
I'm not sure what to make of these logs.  Since you are planing to use MessagePack later though, can you drop us a bug report via the Unity editor? Then we can investigate more from our side.

**Comment by anurse** [comment]
> would you be able to point me to page to track some information when 3.0.0 will be considered to be planned

This repo, and aspnet/Announcements is where most of our planning discussions and decisions are announced. We're currently working on shipping 2.2 by the end of this year (see the [ASP.NET Core 2.2 Roadmap](https://github.com/aspnet/Announcements/issues/307) for more information). 3.0 will be the release that follows 2.2.

**Comment by dzmitry-lahoda** [comment]
Possible Unity side reported here https://fogbugz.unity3d.com/default.asp?1091189_1sqkebcrot7vvv9b

**Comment by dzmitry-lahoda** [comment]
I did it https://github.com/neuecc/MessagePack-CSharp/issues/338 . So it seems need some changes into MessagePack-CSharp to be `efficient` and probably Unity should fix Expression interpreter to allow efficient, but working solution from start.

**Comment by anurse** [comment]
As this is a limitation of the MessagePack-CSharp library, I'm going to close this for now. Let us know if there are specific things we can help drive to get MessagePack working in AOT environments. We'd like to get things working there :).

**Comment by dzmitry-lahoda** [comment]
1. SignalR MessagePack integration does calls routines of MessagePack which do IL generation even if non of generated code is used by SignalR or my code later. That happens during SignalR builder construction. I have shown fix of SignalR integration earlier. Fix should be on SignalR integration with MessagePack side. Not MessagePack. Types of MessagePack which do IL generation should not be touched by builder if I provide formatters manually.

2. Fixed first, going next. MessagePack tries to evaluate LINQ expression via interpreter. Interpreter fails. Unity support for .NET Standard 2.0 seems to be issue here. They should reject LINQ interpenetration totally with proper error or fully support, but not null reference or index out of bound, what ever. 

3. I workaround that either by avoiding LINQ via manual calls of serializer, also in SignalR integration. So I fixed that issue in SignalR integration, not in MessagePack. But that issue seems to be better fixed by MessagePack library by avoding LINQ totally. Using Delegate.Create, or MPC do generate expression, or cached reflection invoke will work.

I say here that each party has issues to fix.
